### PR TITLE
[Feature] support mIoU metric

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

Support evaluating semantic segmentation task

## Modification

Add mIoU metric

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/evaluation/metrics/miou_metric.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+                    data_sample['img_path']))[0]
+                png_filename = osp.abspath(
+                    osp.join(self.output_dir, f'{basename}.png'))
+                output_mask = pred_label.cpu().numpy()
+                # The index range of official ADE20k dataset is from 0 to 150.
+                # But the index range of output is from 0 to 149.
+                # That is because we set reduce_zero_label=True.
+                if data_sample.get('reduce_zero_label', False):
+                    output_mask = output_mask + 1
+                output = Image.fromarray(output_mask.astype(np.uint8))
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:99]: 不应该直接用save，而应该用  mmcv.imwrite，因为他可以支持存储到不同后端 backend_args

File: mmdet/evaluation/metrics/miou_metric.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+            Returns:
+                [torch.tensor]: The f-score value.
+            """
+            score = (1 + beta**2) * (precision * recall) / (
+                (beta**2 * precision) + recall)
+            return score
+
+        if isinstance(metrics, str):
+            metrics = [metrics]
+        allowed_metrics = ['mIoU', 'mDice', 'mFscore']
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:249]: 这个应该放一开始，不然会出现用户跑到最后一张图报这个错误，用户会喷

File: mmdet/evaluation/metrics/miou_metric.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        if not set(metrics).issubset(set(allowed_metrics)):
+            raise KeyError(f'metrics {metrics} is not supported')
+
+        all_acc = total_area_intersect.sum() / total_area_label.sum()
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:253]: all_acc 还是 avg_acc?

File: mmdet/evaluation/metrics/miou_metric.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        ret_metrics = OrderedDict({'aAcc': all_acc})
+        for metric in metrics:
+            if metric == 'mIoU':
+                iou = total_area_intersect / total_area_union
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:257]: 有做边界处理吗？total_area_union 会有 0 的情况吗？

File: mmdet/evaluation/metrics/miou_metric.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            elif metric == 'mDice':
+                dice = 2 * total_area_intersect / (
+                    total_area_pred_label + total_area_label)
+                acc = total_area_intersect / total_area_label
+                ret_metrics['Dice'] = dice
+                ret_metrics['Acc'] = acc
+            elif metric == 'mFscore':
+                precision = total_area_intersect / total_area_pred_label
+                recall = total_area_intersect / total_area_label
+                f_value = torch.tensor([
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:270]: beta 好像没有用 self.beta?

File: mmdet/evaluation/metrics/miou_metric.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        """
+        num_classes = len(self.dataset_meta['classes'])
+        for data_sample in data_samples:
+            pred_label = data_sample['pred_sem_seg']['data'].squeeze()
+            # format_only always for test dataset without ground truth
+            if not self.format_only:
+                label = data_sample['gt_sem_seg']['sem_seg'].squeeze().to(
+                    pred_label)
+                self.results.append(
+                    self.intersect_and_union(pred_label, label, num_classes,
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:85]: intersect_and_union 这个名字不合适，应该要通用点。假设后面要扩展一个新的metric，就需要修改这个函数，那么就不合适了

File: mmdet/evaluation/metrics/miou_metric.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+            intersect.float(), bins=(num_classes), min=0,
+            max=num_classes - 1).cpu()
+        area_pred_label = torch.histc(
+            pred_label.float(), bins=(num_classes), min=0,
+            max=num_classes - 1).cpu()
+        area_label = torch.histc(
+            label.float(), bins=(num_classes), min=0,
+            max=num_classes - 1).cpu()
+        area_union = area_pred_label + area_label - area_intersect
+        return area_intersect, area_union, area_pred_label, area_label
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:200]: 为了通用性，返回最好是 dict，而非 list，并且 .cpu 的计算应该不是必须的，应该要和 collect_device 绑定。如果是 cuda，其实就最好不要转，没有必要。

File: mmdet/evaluation/metrics/miou_metric.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+    @staticmethod
+    def total_area_to_metrics(total_area_intersect: np.ndarray,
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:203]: 输入应该就是前面返回的dict 就行，函数名也换一下，通用些

Comment (hhaAndroid) [reply mmdet/evaluation/metrics/miou_metric.py:203]: 类型可能不是 np

File: mmdet/evaluation/metrics/miou_metric.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+                              total_area_union: np.ndarray,
+                              total_area_pred_label: np.ndarray,
+                              total_area_label: np.ndarray,
+                              metrics: List[str] = ['mIoU'],
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/miou_metric.py:207]: 这个参数不用传，这个函数也不用设置为静态方法吧

**Review by hhaAndroid — APPROVED** [review]
